### PR TITLE
refactor(forms): use transient props where appropriate

### PR DESCRIPTION
**Pull request by ze-flo**
## Description
This PR updates various components in `react-forms` to use [transient props](https://styled-components.com/docs/api#transient-props) where appropriate. These changes are necessary in  preparation for the upgrade to `styled-components@6.x.x` to ensure we avoid DOM violation errors after the transition.

## Checklist

- [ ] ~~:ok_hand: design updates will be Garden Designer approved (add the designer as a reviewer)~~
- [x] :globe_with_meridians: demo is up-to-date (`npm start`)
- [ ] ~~:arrow_left: renders as expected with reversed (RTL) direction~~
- [ ] ~~:black_circle: renders as expected in dark mode~~
- [ ] ~~:metal: renders as expected with Bedrock CSS (`?bedrock`)~~
- [x] :guardsman: includes new unit tests. Maintain existing coverage (always >= 96%)
- [ ] ~~:wheelchair: tested for WCAG 2.1 AA accessibility compliance~~
- [ ] ~~:memo: tested in Chrome, Firefox, Safari, and Edge~~

### CODE REVIEW

File: packages/forms/src/elements/common/Legend.tsx [hunk 1]
```diff
@@ -7,13 +7,10 @@
 
 import React, { forwardRef, HTMLAttributes } from 'react';
 import { StyledLegend } from '../../styled';
-import useFieldsetContext from '../../utils/useFieldsetContext';
 
 const LegendComponent = forwardRef<HTMLLegendElement, HTMLAttributes<HTMLLegendElement>>(
   (props, ref) => {
-    const fieldsetContext = useFieldsetContext();
```
Comment (ze-flo) [inline packages/forms/src/elements/common/Legend.tsx:14]: `StyledLegend` extends `StyledLabel`, which [doesn't support](https://github.com/zendeskgarden/react-components/blob/main/packages/forms/src/styled/common/StyledLabel.ts) `isCompact`. Removing.

File: packages/forms/src/styled/common/StyledMessageIcon.ts [hunk 2]
```diff
@@ -35,8 +35,12 @@ const MessageIcon: React.FC<PropsWithChildren<IStyledMessageIconProps>> = ({
 
 const COMPONENT_ID = 'forms.input_message_icon';
 
+/**
+ * 1. Cannnot use transient prop `$validation`.
+ * MessageIcon is not a styled component and will not receive the prop.
+ */
```
Comment (ze-flo) [inline packages/forms/src/styled/common/StyledMessageIcon.ts:41]: FYI: Styled-components only forwards transient props to components wrapped in `styled`. [MessageIcon](https://github.com/zendeskgarden/react-components/pull/1952/files#diff-9bb2af49ef566d88c576332118c64b069cc91d2294523e771344ae10983e1d65R16-R34) is a traditional FC and wouldn't receive props prefixed with `$`.

**Comment by ze-flo** [comment]
All snapshots tests in #1953 have passed. ✅

**Review by jzempel — APPROVED** [review]

File: packages/forms/src/styled/common/StyledMessageIcon.ts (hunk 2 above)
Comment (ze-flo) [reply packages/forms/src/styled/common/StyledMessageIcon.ts]: Fixed after pulling changes from #1954. 🙌

**Comment by ze-flo** [comment]
Snapshots are [failing](https://github.com/zendeskgarden/react-components/actions/runs/11373805698/job/31641495729) due to this [change](https://github.com/zendeskgarden/react-components/commit/30e06b0ae8b41de67d70bad897390ff8409aaa29#diff-927f3bc9a332a05dbdc2302b5585fa3703c4fa41ced1622aa8d51d58d0661263R66). This is expected and should be okay to merge.

**Pull request merged**